rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,586 @@
+                        "Parquet datasource does not support LIST".to_string(),
+                    ));
+                }
+                DataType::Struct(_) => {
+                    return Err(ExecutionError::NotImplemented(
+                        "Parquet datasource does not support STRUCT".to_string(),
+                    ));
+                }
+                _ => {}
+            }
+        }
+
+        let projection = match projection {
+            Some(p) => p,
+            None => {
+                let mut p = Vec::with_capacity(schema.fields().len());
+                for i in 0..schema.fields().len() {
+                    p.push(i);
+                }
+                p
+            }
+        };
+
+        let projected_schema = schema_projection(&schema, &projection)?;
+
+        Ok(ParquetFile {
+            reader: reader,
+            row_group_index: 0,
+            projection_schema: projected_schema,
+            projection,
+            batch_size: 64 * 1024,
+            current_row_group: None,
+            column_readers: vec![],
+        })
+    }
+
+    fn load_next_row_group(&mut self) -> Result<()> {
+        if self.row_group_index < self.reader.num_row_groups() {
+            let reader = self.reader.get_row_group(self.row_group_index)?;
+
+            self.column_readers = Vec::with_capacity(self.projection.len());
+
+            for i in 0..self.projection.len() {
+                self.column_readers
+                    .push(reader.get_column_reader(self.projection[i])?);
+            }
+
+            self.current_row_group = Some(reader);
+            self.row_group_index += 1;
+
+            Ok(())
+        } else {
+            Err(ExecutionError::General(
+                "Attempt to read past final row group".to_string(),
+            ))
+        }
+    }
+
+    fn load_batch(&mut self) -> Result<Option<RecordBatch>> {
+        match &self.current_row_group {
+            Some(reader) => {
+                let mut batch: Vec<Arc<Array>> = Vec::with_capacity(reader.num_columns());
+                for i in 0..self.column_readers.len() {
+                    let is_nullable = self.schema().field(i).is_nullable();
+                    let array: Arc<Array> = match self.column_readers[i] {
+                        ColumnReader::BoolColumnReader(ref mut r) => {
+                            ArrowReader::<BooleanType>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::Int32ColumnReader(ref mut r) => {
+                            ArrowReader::<Int32Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::Int64ColumnReader(ref mut r) => {
```

Comment:
should we look at the logical type for the `int64`? it may need to be converted to timestamp or decimal. Same for `int32`.

Reply:
Yes, that's a good point.

Reply:
Thanks. It's fine to do it in a separate PR though.

Reply:
How would you feel about doing this as separate PRs after this is merged?

Reply:
Yes. That's fine for me.

Reply:
Cool. Once you and/or Paddy have approved this one, I'll merge and start on the date/time support.